### PR TITLE
layout: Port `border-image` support from legacy layout

### PR DESCRIPTION
**Pull request by mrobinson**
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

### CODE REVIEW

**Comment by sagudev** [comment]
> Port border-image support for legacy layout

Did you mean `Port border-image support from legacy layout`?

**Comment by mrobinson** [comment]
> > Port border-image support for legacy layout
> 
> Did you mean `Port border-image support from legacy layout`?

Yes

File: components/layout_2020/display_list/mod.rs [hunk 1]
```diff
@@ -1118,3 +1257,72 @@ fn create_clip_chain(
...
+        image_width_for_side(&width.0, border.top, border_area.height),
+        image_width_for_side(&width.1, border.right, border_area.width),
+        image_width_for_side(&width.2, border.bottom, border_area.height),
+        image_width_for_side(&width.3, border.left, border_area.width),
+    )
+}
+
+/// Get the WebRender border-image slice from the style values.
+///
+/// FIXME(emilio): WR expects device pixels here... somehow?
```
Comment (mukilan) [inline components/layout_2020/display_list/mod.rs:1310]:
```suggestion
/// FIXME: WR expects device pixels here... somehow?
```
Is this FIXME still relevant?

File: components/layout_2020/display_list/mod.rs [hunk 2]
```diff
@@ -1118,3 +1257,72 @@ fn create_clip_chain(
             .define_clip_chain(parent_clip_chain_id, [new_clip_id]),
     )
 }
+
+/// Compute the additional border-image area.
+pub fn resolve_border_image_outset(
```
Comment (mukilan) [inline components/layout_2020/display_list/mod.rs:1262]: Do these new `resolve_*` functions need to be `pub`?

**Review by mukilan — APPROVED** [review]

File: components/layout_2020/display_list/mod.rs (hunk 1 above)
Comment (mrobinson) [reply components/layout_2020/display_list/mod.rs:1310]: I guess not. It sort of makes sense that WebRender takes device pixels here as the values measure pixels of the raster image.

File: components/layout_2020/display_list/mod.rs (hunk 2 above)
Comment (mrobinson) [reply components/layout_2020/display_list/mod.rs:1262]: Nope! I've made them private and have also added more documentation here.

**Pull request merged**